ruby/ql/src/experimental/manually-check-http-verb/ManuallyCheckHttpVerb.ql

```diff
@@ -0,0 +1,58 @@
+/**
+ * @name Manually checking http verb instead of using built in rails routes and protections
+ * @description Manually checking HTTP verbs is an indication that multiple requests are routed to the same controller action. This could lead to bypassing necessary authorization methods and other protections, like CSRF protection. Prefer using different controller actions for each HTTP method and relying Rails routing to handle mappting resources and verbs to specific methods.
+ * @kind problem
+ * @problem.severity error
+ * @security-severity 5.0
+ * @precision low
+ * @id rb/manually-checking-http-verb
+ * @tags security
+ */
+
+import ruby
+import codeql.ruby.DataFlow
+
+class CheckNotGetRequest extends ConditionalExpr {
+  CheckNotGetRequest() { this.getCondition() instanceof CheckGetRequest }
+}
+
+class CheckGetRequest extends MethodCall {
+  CheckGetRequest() { this.getMethodName() = "get?" }
+}
+
+class ControllerClass extends ModuleBase {
+  ControllerClass() { this.getModule().getSuperClass+().toString() = "ApplicationController" }
+}
+
+class CheckGetFromEnv extends AstNode {
+  CheckGetFromEnv() {
+    // is this node an instance of `env["REQUEST_METHOD"]
+    this instanceof GetRequestMethodFromEnv and
+    (
+      // and is this node a param of a call to `.include?`
+      exists(MethodCall call | call.getAnArgument() = this | call.getMethodName() = "include?")
+      or
+      // check if env["REQUEST_METHOD"] is compared to GET
+      exists(EqualityOperation eq | eq.getAChild() = this |
+        eq.getAChild().(StringLiteral).toString() = "GET"
+      )
+    )
+  }
```

Comment:
The idea here is that I would also like to find instances where the value of `env["REQUEST_METHOD"] is compared to a string literal as part of a control flow check. I'm having a bit of trouble wrapping my head around how I'd constrain things to be part of a ConditionalExpr from here, or am I thinking about this the wrong way? Should I just add that part directly to my query?

Reply:
Also, does this seem useful or should I just stick to singling out calls to `request.get?`?

Reply:
Ah awesome! I've been leaning on `.toString()` far too much, so this is great!